File: python_code/complete/no008.py
#!/usr/bin/env python

# Find the greatest product of five consecutive
# digits in the 1000-digit number.

import operator

from python_code.decorators import euler_timer
from python_code.functions import get_data

def product_consec_digits(number, consecutive):
    """
    Returns the largest product of "consecutive"
    consecutive digits from number
    """
    digits = [int(dig) for dig in str(number)]
    max_start = len(digits) - consecutive
    return [reduce(operator.mul, digits[i:i + consecutive])
            for i in range(max_start + 1)]

def main(verbose=False):
    n = int("".join([line.strip() for line in get_data(8).split("\n")]))

    return max(product_consec_digits(n, 5))

if __name__ == "__main__":
    print euler_timer(8)(main)(verbose=True)

File: js/complete/no008.js
#!/usr/bin/env node

/* Find the greatest product of five consecutive
   digits in the 1000-digit number. */

var fns = require('../functions.js'),
    operator = require('../operator.js'),
    timer = require('../timer.js');

function productConsecDigits(number, consecutive) {
    /* Returns the largest product of 'consecutive'
       consecutive digits from number */
    var digits = number.toString().split(''),
        maxStart = digits.length - consecutive,
        result = [];

    /* Could track the product instead of reducing each time.
       Just doing what python does. */
    for (var i = 0; i < maxStart + 1; i++) {
	result.push(digits.slice(i, i + consecutive).reduce(operator.mul, 1));
    }

    return result;
};

exports.main = function(verbose) {
    if (typeof verbose == 'undefined') {
        verbose = false;
    }
    var data = fns.getData(8).split('\n').join(''),
        products = productConsecDigits(data, 5);
    return Math.max.apply(Math, products);
};

if (require.main === module) {
    timer.timer(8, exports.main);
}
